design-proposals/klip-4-produce-data.md

```diff
@@ -0,0 +1,277 @@
+in KSQL (e.g. as a result of `CREATE STREAM`)
+
+## Value
+
+This proposal will improve the first 15-minute experience that developers have with KSQL and will
+help seasoned KSQL developers test new functionality. Furthermore, this functionality can be
+leveraged to create a powerful testing harness for KSQL applications that can run in a sandbox
+without pre-existing data.
+
+## Public APIS
+
+#### INSERT INTO
+
+```sql
+INSERT INTO (stream_name|table_name) (column1, column2, ...)
+  VALUES (value1, value2, ...)
+```
+
+The `INSERT INTO` statement can be used to insert new records into a kafka topic (agnostic to the
+Stream/Table duality). If the columns are not present, it is assumed that the values are in the same
+order as the schema and contain every field. The key for the insert statement is taken from the
+column the matches the name of `key` in the `CREATE` statement properties for the corresponding
+source. This functionality will need to be extended when we support structured keys.
+
+The value for `stream_name`/`table_name` must be a valid Stream/Table registered with a KSQL schema.
+The serialization format will be the same as the format specified in the `value_format` of the
+`CREATE` statement. 
+
+For the initial version, only primitive literal types will be supported (e.g. `INT`, `VARCHAR`, 
+etc...) but syntax is easily extensible to support struct, map and list (i.e. by following syntax 
+like Hive's `NAMED_STRCUT` construct). 
+
+Future extensions: 
+* Support some`AT INTERVAL` syntax to generalize the functionality to generate full stream data.
+* Read from file to allow reproducible insertions
+
+#### DELETE FROM
+
+```sql 
+DELETE FROM table_name WHERE condition;
+```
+
+The `DELETE` statement can be used to insert tombstones into kafka topics that back tables. If the
+source is not back a table, the statement will fail. For this KLIP, we will only support conditions
+of the form `ROWKEY = value`. While this may initially be confusing (as SQL standard allows deleting
+on any arbitrary condition), we have precedent for this in stream-table joins (the join criteria
+may only be on the key of the table).
+
+#### CREATE STREAM/TABLE
+
+```sql
+CREATE STREAM stream_name ( { column_name data_type } [, ...] )
+  WITH ( property_name = expression [, ...] );
+```
+
+```sql
+CREATE TABLE table_name ( { column_name data_type } [, ...] )
+  WITH ( property_name = expression [, ...] );
+```
+
+The `WITH` clause will now accept two additional property_names: `PARTITIONS`/`REPLICAS` If the 
+topic *does not exist* and these values are specified, this command will create a topic with a the
+set number of partitions and replicas before executing the command. If the topic exists and has
+a different number of partitions/replicas, this command will fail.
+
+This functionality will be flagged under a configuration `ksql.ddl.source.topic.create.allow`,
```

Comment:
Copying comment from discussion above: 

> I do believe, though, that we should make it explicit for CT/CS statements that you intend to create an empty topic - this will help catch typos if you intended to create a stream/table off of an existing topic (we don't want that type of behavior enabled for "production" use cases and definitely not for headless mode). I think a global config is sufficient here (instead of having it configured per statement) because it seems that you would either want all CT/CS statements to optionally create topics (testing), or none of them (production).

Reply:
Hummmm....

So the purpose of the config is to allow users, (admins), to configure the cluster to not allow arbitrary topics via CT/CS statements... but, as Robin says, you can already do that via CSAS/CTAS statements.  

I'm also not sold on the whole 'this will only be turned on in testing and never in production' thing.   Who knows how people will use this!  One of the big 'issues' with streaming computation is that it's often pretty tricky to correct an error / discrepancy.   `INSERT VALUES` could be just what some 'Streaming DBAs' need to be able to correct production data when the CFOs shouting at them at 3 in the morning.

Assuming the Kafka cluster is locked down with ACLs then its the ACLs that will define if a user has permissions to create a topic or not or if they have permissions to insert into, i.e. produce, into a topic or not. 

So... I'm kind of leaning towards Robin's view that such a config would be of limited use...

Reply:
This seems convincing enough. I think the flag I really want is the one at the bottom of the KLIP (`ksql.insert.into.values.enabled`), I'm happy removing this flag.

Reply:
>One of the big 'issues' with streaming computation is that it's often pretty tricky to correct an error / discrepancy. INSERT VALUES could be just what some 'Streaming DBAs' need to be able to correct production data when the CFOs shouting at them at 3 in the morning.

This is exactly the use case I was describing when talking to @hjafarpour  😂